### PR TITLE
Implement Sparse Face Detection Model

### PR DESCRIPTION
**Pull request by ShikharJ**

### CODE REVIEW

File: c_reference/include/quantized_fastgrnn.h [hunk 1]
```diff
@@ -183,7 +187,11 @@ typedef struct Q7xQ15_FastGRNN_Params {
   const Q7_T* mean;
   const Q7_T* stdDev;
   const Q15_T* W;
+  const ITER_T* Wids;
+  const Q15_T* Wvals;
```
Comment (harsha-simhadri) [inline c_reference/include/quantized_fastgrnn.h:191]: why are we adding more variables to fastgrnn params struct. would it be better to start another struct for sparse fastgrnn?

File: c_reference/src/quantized_fastgrnn.c [hunk 2]
```diff
@@ -204,12 +213,21 @@ int q15_fastgrnn(Q15_T* const hiddenState, ITER_T hiddenDims,
     }
 
     // Process the new input and previous hidden state
-    q15_m_mulvec(tparams->W, tbuffers->normFeatures, hiddenDims, inputDims,
-      tbuffers->preComp1, tscales->W, tscales->normFeaturesMVW, tscales->H1W,
-      tscales->H2W);
-    q15_m_mulvec(tparams->U, hiddenState, hiddenDims, hiddenDims,
-      tbuffers->preComp2, tscales->U, tscales->hiddenStateMVU, tscales->H1U,
-      tscales->H2U);
+    #ifdef SPARSE
```
Comment (harsha-simhadri) [inline c_reference/src/quantized_fastgrnn.c:216]: should we interleave sparse code like this or separate out to another set of files?

**Review by harsha-simhadri — COMMENTED** [review]
What are the merits of folding the sparse support into the dense struct and functions instead of creating a new set of structs and function calls?

File: c_reference/src/quantized_fastgrnn.c (hunk 2 above)
Comment (ShikharJ) [reply c_reference/src/quantized_fastgrnn.c:216]: As far as `RNNPool` is concerned, these two calls are the only two sparsity related changes needed. Separating out to another set of files would be wasteful in my opinion.

File: c_reference/include/quantized_fastgrnn.h (hunk 1 above)
Comment (ShikharJ) [reply c_reference/include/quantized_fastgrnn.h:191]: If we add another `struct` for sparse params, the `fastgrnn()` function will have to be duplicated again, just for changing one line of code:
```
const Q7xQ15_FastGRNN_Params* tparams = (const Q7xQ15_FastGRNN_Params*)params;
```
Hence, I'm avoiding this for now. Folding sparse support into our existing dense framework saves us a lot of duplication.

Comment (harsha-simhadri) [reply c_reference/include/quantized_fastgrnn.h:191]: OK, do we want to leave some comments in the struct header file so people using it in dense mode dont worry about these four additional params?

**Comment by ShikharJ** [comment]
@harsha-simhadri This is updated now.

**Review by harsha-simhadri — APPROVED** [review]

**Pull request merged**